# Lecture 5_1: Stationary Distributions of CTDS processes

In [4]:
import numpy as np
import math
import matplotlib.pyplot as plt
import random
from scipy.special import comb

****
<span style="color:red;">**Example: 5.4** Species coexistence  in continuous time cont</span>

**Using the transition rate matrix above:**
$$
\mathbf{Q}=\begin{bmatrix} 
-c_B & c_B & 0\\ 
e_B &-(e_B+e_A) & e_A\\
0 & c_A & - c_A\\
\end{bmatrix}
$$
**and letting** 
$$ 
c_B=0.3\frac{1}{\text{year}}\quad c_A=0.1\frac{1}{\text{year}}\quad e_B=0.25\frac{1}{\text{year}}\quad e_A=0.01\frac{1}{\text{year}}
$$
**1. What is the probability of going from state 2 to state 1 in the long term?**

In [8]:
QMtrx=np.array([[-0.3,0.3,0],[0.25,-0.26,0.01],[0,0.1,-0.1]])

In [9]:
eigenvalues, eigenvectors = np.linalg.eig(QMtrx)

In [33]:
np.round(eigenvalues,4)

array([-0.5556, -0.    , -0.1044])

The second eigenvalues is the leading eigenvalue

In [21]:
A=eigenvectors
D=np.diag(eigenvalues)
AInv=np.linalg.inv(A)

Double checking our decomposition

In [23]:
np.round(A@D@AInv,4)

array([[-0.3 ,  0.3 , -0.  ],
       [ 0.25, -0.26,  0.01],
       [ 0.  ,  0.1 , -0.1 ]])

The eigenvectors

In [59]:
u1=A[:,1]
## Normalizing the vector so EACH value is 1
u1=A[:,1]/A[1,1]
print(u1)

[1. 1. 1.]


In [60]:
## Normalizing the vector so the values sum to 1
v1=AInv[1]/np.sum(AInv[1])
print(v1)

[0.43103448 0.51724138 0.05172414]


Calculating $\mathbf{P}$ (note that due to the normalization above each row sums to 1 as it should for a transition probability matrix)

In [61]:
np.outer(u1,v1)

array([[0.43103448, 0.51724138, 0.05172414],
       [0.43103448, 0.51724138, 0.05172414],
       [0.43103448, 0.51724138, 0.05172414]])

The probability of going from state 2 to state 1 over a very long time is $0.43$

**2. What is the long-term probability of ending up in state 1?**

Note that each row is the same, this is because in the long term, the starting condition doesn't matter.  This implies that each row (which is given by $\vec{v}_1$) is the stationary distribution.  Hence, the long-term probability of ending up in state 1 is also $0.43$.
****